suite/tests/client-interface/drx-scattergather.dll.c

```diff
@@ -146,8 +145,28 @@ event_bb_app2app(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
             if (next_instr != NULL) {
                 if (instr_is_mov_constant(next_instr, &val) &&
                     val == TEST_MASK_CLOBBER_MARKER) {
-                    one_time_mask_clobber_test = true;
-                    one_time_mask_clobber_translating = true;
+                    byte *pc = instr_get_app_pc(next_instr);
+                    instr_t temp_instr;
+                    instr_init(drcontext, &temp_instr);
+                    /* We're searching for the next gather instruction that will be
+                     * expanded below. We will use its pc to identify the corner case
+                     * instructions where we will inject a ud2 after gather expansion.
+                     * This relies heavily on the exact test app's behavior, as well as
+                     * the scatter/gather expansion's code layout.
+                     */
+                    while (true) {
+                        instr_reset(drcontext, &temp_instr);
+                        byte *next_pc = decode(drcontext, pc, &temp_instr);
+                        CHECK(pc != NULL,
+                              "Everything should be decodable in the test until a "
+                              "gather instruction will be found.");
+                        if (instr_is_gather(&temp_instr)) {
+                            mask_clobber_test_gather_pc = pc;
+                            break;
+                        }
+                        pc = next_pc;
```

Comment:
I would assert we don't hit a cti, which will catch false positive pattern matches

Reply:
Done.